app/javascript/hyacinth_ui_v1/components/projects/permissions/PermissionsEditor.js

```diff
@@ -0,0 +1,300 @@
+        }
+      }),
+    );
+  };
+
+  const removeUser = (userId) => {
+    setPermissionsChanged(true);
+    setProjectPermissions(
+      produce(projectPermissions, (draft) => {
+        arrRemove(draft, (projectPermission) => {
+          return projectPermission.user.id === userId;
+        });
+      }),
+    );
+    setRemovedUserIds(
+      produce(removedUserIds, (draft) => {
+        draft.add(userId);
+      }),
+    );
+  };
+
+  const renderProjectPermission = (projectPermission) => {
+    const {
+      projectPermissionActions: {
+        actions, readObjectsAction, actionsDisallowedForAggregatorProjects,
+      },
+    } = actionsData;
+
+    return actions.map((action) => {
+      const disabledCheckbox = readOnly || action === readObjectsAction || (
+        !projectPermission.project.isPrimary
+        && actionsDisallowedForAggregatorProjects.includes(action)
+      );
+
+      return (
+        <td key={action}>
+          { /* eslint-disable jsx-a11y/label-has-associated-control, jsx-a11y/label-has-for */ }
+          { /* Note 1: ESLint is getting confused by label + react-bootstrap element. */ }
+          { /* Note 2: We're Using a label for a wider clickable checkbox area. */ }
+          <label
+            id={`label-${projectPermission.user.id}-${action}`}
+            htmlFor={`checkbox-${projectPermission.user.id}-${action}`}
+            className="w-100"
+            >
+            <Form.Check
+              id={`checkbox-${projectPermission.user.id}-${action}`}
+              label="&nbsp;"
+              disabled={disabledCheckbox}
+              type="checkbox"
+              checked={projectPermission.actions.includes(action)}
+              onChange={(e) => {
+                if (readOnly) return;
+                updatePermissionsData(projectPermission.user.id, action, e.target.checked);
+              }}
+            />
+          </label>
+          { /* eslint-enable jsx-a11y/label-has-associated-control, jsx-a11y/label-has-for */ }
+        </td>
+      );
+    });
+  };
+
+  const renderProjectPermissions = () => {
+    return projectPermissions.map(projectPermission => (
+      <tr key={projectPermission.user.id}>
+        <td key="name">
+          {projectPermission.user.fullName}
+        </td>
+
+        {renderProjectPermission(projectPermission)}
+
+        { !readOnly && (
+          <td key="remove" className="text-center">
+            <RemoveButton onClick={() => {removeUser(projectPermission.user.id); }} />
+          </td>
+        ) }
+      </tr>
+    ));
+  };
+
+  const addSelectedUser = () => {
+    if (currentAddUserSelection == null) return;
+    setPermissionsChanged(true);
+
+    const { projectPermissionActions: { readObjectsAction } } = actionsData;
+    const userId = currentAddUserSelection.value;
+    const user = allUsers.find(u => u.id === userId);
+
+    setProjectPermissions(produce(projectPermissions, (draft) => {
```

Comment:
I was wondering if I could do something like that.  Question though: is it okay to reuse all of the other existing non-primitive-value elements in the array when setting state, or should I treat all of those as immutable and dup them for that reason?  That was where this choice came from, but maybe I'm being way more cautious than I need to be.

Is reusing read-only nested elements when setting state always okay?

Reply:
I didn't think it was necessary, but I might be terribly wrong about that. I'll do some research. I still think we should explore a custom hook for `setProjectPermissions` tho.

Reply:
Thanks!  And I'm interested in the custom hook route!

Reply:
I think you're right that for complicated state changes we should use immer (or other means) to make sure the state is entirely copied. Though I can't find sources that tell me this specifically. I guess since I have only used hooks for very simple state, one primative or maybe an array of primitive values I haven't really needed to use immer. I am looking into the custom hook now.

Reply:
I'll work on the custom hook as a separate PR. I think it would be very helpful to have the custom hook when we create the User interface for project permissions.